src/transformers/integrations/flex_attention.py

```diff
@@ -1,43 +1,185 @@
+            self._compiled_flex_attention = torch.compile(flex_attention, dynamic=False)
+            self._is_flex_compiled = True
+
+    def __call__(self):
+        return self._compiled_flex_attention
+
+
+def make_flex_block_causal_mask(attention_mask_2d: torch.Tensor) -> BlockMask:
+    """
+    Create a block causal document mask for a batch of sequences, both packed and unpacked.
+    Create Block causal logic and passing it into :func:`torch.nn.attention.flex_attention.create_block_mask`.
+    The resultant BlockMask is a compressed representation of the full block causal
+    mask. BlockMask is essential for performant computation of flex attention.
+    See: https://pytorch.org/blog/flexattention/
+
+    Args:
+        attention_mask_2d (torch.Tensor): Attention mask for packed and padded sequences
+        of shape (batch_size, total_seq_len). e.g.
+
+        For unpacked sequence:
+        [[1, 1, 1, 1, 0, 0, 0],
+         [1, 1, 1, 1, 1, 0, 0]]
+
+        For packed sequence:
+        [[1, 1, 1, 2, 2, 2, 0],
+         [1, 1, 2, 2, 2, 3, 3]]
+
+    Returns:
+        BlockMask
+    """
+    device = attention_mask_2d.device
+
+    document_ids = attention_mask_2d
+    batch_size, total_seq_len = document_ids.shape
+
+    # Instead of passing a tensor mask, flex attention requires a mask_mod function
+    # that determines which elements of QK^T should be included in the attention
+    # computation prior to the softmax. For sample packing, we need both the
+    # logic for both causal mask and document mask. See PyTorch's official
+    # blog post for more details: https://pytorch.org/blog/flexattention/#mask-mods
+    def causal_mask_mod(batch_idx, head_idx, q_idx, kv_idx):
+        """
+        Defines the logic of a block causal mask by combining both a standard causal mask
+        and a block diagonal document mask.
+
+        See :func:`~torchtune.modules.attention_utils.create_block_causal_mask`
+        for an illustration.
+        """
+        causal_mask = q_idx >= kv_idx
+        document_mask = document_ids[batch_idx, q_idx] == document_ids[batch_idx, kv_idx]
+        padding_mask = document_ids[batch_idx, q_idx] > 0
+        return causal_mask & document_mask & padding_mask
+
+    return create_block_causal_mask_flex(
+        mask_mod=causal_mask_mod,
+        B=batch_size,
+        H=None,  # attention head
+        Q_LEN=total_seq_len,
+        KV_LEN=total_seq_len,
+        device=device,
+    )
+
+
+@torch.compiler.disable(recursive=False)
+def compile_friendly_flex_attention(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    **kwargs,
+) -> torch.Tensor:
+    # First call initialise singleton wrapper object, second call invokes the object method to return compiled flex attention
+    flex_attention_compiled = WrappedFlexAttention()()
+    return flex_attention_compiled(
+        query,
+        key,
+        value,
+        **kwargs,
+    )
```

Comment:
thanks for your great work! just curious, why don't we just return a pre-compiled flex_attention (just like torchtune https://github.com/pytorch/torchtune/blob/main/torchtune/modules/attention_utils.py#L44-L57 ) ？